### PR TITLE
Update recover accounts

### LINKED ISSUE #1425: Add an `accountStartIndex` to `AccountSyncOptions`
## Description

Balance finding algorithms could be improved if we can specify a starting index for the accounts in addition to the addresses. If we can only adjust the starting address index then it's possible that any new accounts searched will not start from address index zero, which may result in funds not found.

Another improvement consists of using the gap limit `j` as follows:

0) Set address range to equal the gap limit  (`range = [0, j]`)
1) Look through the address range for addresses with funds . 
2) Two scenarios are posibble:
 - If no address is found go to step 3. 
 - If an address is found at `i < j`. Then update the address range to `[i, i + j]` and go to Step 1.
3) Stop searching

## Motivation

See description.

## Requirements

Write a list of what you want this feature to do.

1. Allow for a `accountStartIndex` parameter in the `AccountSyncOptions`
2. Allow both the `accountStartIndex` and `addressStartIndex` to override the start in the case where the start index is lower than where funds were actually found, e.g. if wallet-rs knows funds reside on address 10 but we specify to start at address 5 then it should start at address 5

## Are you planning to do it yourself in a pull request?

Maybe (@Tuditi and I would like to try and do this if not too difficult).

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Update recover accounts

## Links to any relevant issues

Fixes https://github.com/iotaledger/wallet.rs/issues/1425

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Running recover accounts example

## Change checklist

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by samuel-rufi — DISMISSED** [review]

File: src/message_interface/message.rs [hunk 1]
```diff
@@ -78,17 +78,18 @@ pub enum Message {
...
         #[serde(rename = "accountGapLimit")]
-        /// Defines how many accounts without unspent outputs will be
-        /// checked, if an account has unspent outputs, the counter is reset
+        /// The number of accounts to search for, after the last account with unspent outputs.
         account_gap_limit: u32,
         #[serde(rename = "addressGapLimit")]
-        /// Defines how many addresses without unspent outputs will be checked in each account, if an
-        /// address has unspent outputs, the counter is reset
+        /// The number of addresses to search for, after the last address with unspent outputs, in
+        ///   each account.
```
Comment (thibault-martinez) [inline src/message_interface/message.rs]:
```suggestion
        /// each account.
```

File: src/message_interface/message.rs [hunk 2]
```diff
@@ -78,17 +78,18 @@ pub enum Message {
...
-        /// Defines how many addresses without unspent outputs will be checked in each account, if an
-        /// address has unspent outputs, the counter is reset
+        /// The number of addresses to search for, after the last address with unspent outputs, in
+        /// each account.
         address_gap_limit: u32,
-        /// SyncOptions to be used during the account recovery process.
-        /// address_start_index and force_syncing will be overwritten in sync_options to not skip addresses, but also
-        /// don't send duplicated requests
+        /// Optional parameter to specify the sync options. The `address_start_index` and `force_syncing`
+        ///   fields will be overwritten to skip existing addresses.
```
Comment (thibault-martinez) [inline src/message_interface/message.rs]:
```suggestion
        /// fields will be overwritten to skip existing addresses.
```

File: src/account/operations/output_finder.rs [hunk 3]
```diff
@@ -8,37 +8,55 @@ use iota_client::secret::GenerateAddressMetadata;
...
+            // outside
+            let (highest_public_address_index, highest_internal_address_index) = {
+                let account = self.read().await;
+                (
+                    account
+                        .public_addresses
+                        .last()
+                        .map(|a| a.key_index)
+                        .expect("account needs to have a public address"),
+                    account.internal_addresses.last().map(|a| a.key_index),
```
Comment (thibault-martinez) [inline src/account/operations/output_finder.rs:54]: Is there no other way at all to do that? Because to full iterations inside an iteration is not idea

Comment (Thoralf-M) [reply src/account/operations/output_finder.rs:54]: What do you mean with full iteration?

Comment (thibault-martinez) [reply src/account/operations/output_finder.rs:54]: Oh this is `Vec::last` and not `Iterator::last`, ignore me 🙈

**Review by thibault-martinez — APPROVED** [review]

**Review by samuel-rufi — APPROVED** [review]

**Pull request merged**